kafka-daq-report/src/main.rs

```diff
@@ -0,0 +1,294 @@
+mod app;
+mod ui;
+
+use anyhow::Result;
+use chrono::{DateTime, Utc};
+use clap::Parser;
+use crossterm::event::{self, DisableMouseCapture, EnableMouseCapture, Event as CEvent, KeyCode};
+use crossterm::execute;
+use crossterm::terminal::{
+    disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen,
+};
+use kagiyama::{AlwaysReady, Watcher};
+use ratatui::{prelude::CrosstermBackend, Terminal};
+use rdkafka::{
+    consumer::{stream_consumer::StreamConsumer, CommitMode, Consumer},
+    message::Message,
+};
+use std::collections::HashMap;
+use std::{
+    io,
+    net::SocketAddr,
+    path::PathBuf,
+    sync::{mpsc, Arc, Mutex},
+    thread,
+    time::{Duration, Instant},
+};
+use streaming_types::dat1_digitizer_analog_trace_v1_generated::{
+    digitizer_analog_trace_message_buffer_has_identifier, root_as_digitizer_analog_trace_message,
+};
+use tokio::task;
+use ui::ui;
+
+use crate::app::App;
+
+/// Holds required data for a specific digitiser.
+pub struct DigitiserData {
+    pub num_msg_received: usize,
+    pub first_msg_timestamp: Option<DateTime<Utc>>,
+    pub last_msg_timestamp: Option<DateTime<Utc>>,
+    pub last_msg_frame: u32,
+    pub num_channels_present: usize,
+    pub has_num_channels_changed: bool,
+    pub num_samples_in_first_channel: usize,
+    pub is_num_samples_identical: bool,
+    pub has_num_samples_changed: bool,
+}
+
+impl DigitiserData {
+    /// Create a new instance with default values.
+    pub fn new(
+        timestamp: Option<DateTime<Utc>>,
+        frame: u32,
+        num_channels_present: usize,
+        num_samples_in_first_channel: usize,
+        is_num_samples_identical: bool,
+    ) -> Self {
+        DigitiserData {
+            num_msg_received: 1,
+            first_msg_timestamp: timestamp,
+            last_msg_timestamp: timestamp,
+            last_msg_frame: frame,
+            num_channels_present,
+            has_num_channels_changed: false,
+            num_samples_in_first_channel,
+            is_num_samples_identical,
+            has_num_samples_changed: false,
+        }
+    }
+}
+
+type SharedData = Arc<Mutex<HashMap<u8, DigitiserData>>>;
+
+#[derive(Debug, Parser)]
+#[clap(author, version, about)]
+struct Cli {
+    #[clap(long)]
+    broker: String,
+
+    #[clap(long)]
+    username: Option<String>,
+
+    #[clap(long)]
+    password: Option<String>,
+
+    #[clap(long = "group")]
+    consumer_group: String,
+
+    #[clap(long)]
+    trace_topic: String,
+
+    #[clap(long, default_value = ".")]
+    output: PathBuf,
+
+    #[clap(long, default_value = "127.0.0.1:9090")]
+    observability_address: SocketAddr,
+}
+
+enum Event<I> {
+    Input(I),
+    Tick,
```

Comment:
Is the `Tick` event used?

Reply:
I copied the event polling algorithm from an online example, and I'm not entirely sure how it works. The TUI does not get displayed if I remove the `Tick` event.

Reply:
Ah, I see how it is being used in the example now.

And it has just occurred to me that the periodic update is actually required here, my suggestion was going to be a refactor that would end up meaning you would have to move the cursor to see any value updating.